core/fetcher.go

```diff
@@ -145,45 +211,133 @@ func (f *BlockFetcher) SubscribeNewBlockEvent(ctx context.Context) (<-chan types
 		defer close(f.doneCh)
 		defer close(signedBlockCh)
 		for {
+			resp, err := subscription.Recv()
+			if err != nil {
+				// case where the context was not canceled but still received an error
+				if ctx.Err() == nil {
+					log.Errorw("fetcher: error receiving new height", "err", err.Error())
+				}
+				return
+			}
+			signedBlock, err := f.GetSignedBlock(ctx, &resp.Height)
+			if err != nil {
+				log.Errorw("fetcher: error receiving signed block", "height", resp.Height, "err", err.Error())
+				return
+			}
 			select {
+			case signedBlockCh <- types.EventDataSignedBlock{
+				Header:       *signedBlock.Header,
+				Commit:       *signedBlock.Commit,
+				ValidatorSet: *signedBlock.ValidatorSet,
+				Data:         *signedBlock.Data,
+			}:
 			case <-ctx.Done():
 				return
-			case newEvent, ok := <-eventChan:
-				if !ok {
-					log.Errorw("fetcher: new blocks subscription channel closed unexpectedly")
-					return
-				}
-				signedBlock := newEvent.Data.(types.EventDataSignedBlock)
-				select {
-				case signedBlockCh <- signedBlock:
-				case <-ctx.Done():
-					return
-				}
 			}
 		}
 	}()
 
 	return signedBlockCh, nil
 }
 
-// UnsubscribeNewBlockEvent stops the subscription to new block events from Core.
-func (f *BlockFetcher) UnsubscribeNewBlockEvent(ctx context.Context) error {
-	f.cancel()
-	select {
-	case <-f.doneCh:
-	case <-ctx.Done():
-		return fmt.Errorf("fetcher: unsubscribe from new block events: %w", ctx.Err())
-	}
-	return f.client.Unsubscribe(ctx, newBlockSubscriber, newDataSignedBlockQuery)
-}
-
 // IsSyncing returns the sync status of the Core connection: true for
 // syncing, and false for already caught up. It can also return an error
 // in the case of a failed status request.
 func (f *BlockFetcher) IsSyncing(ctx context.Context) (bool, error) {
-	resp, err := f.client.Status(ctx)
+	resp, err := f.client.Status(ctx, &coregrpc.StatusRequest{})
 	if err != nil {
 		return false, err
 	}
 	return resp.SyncInfo.CatchingUp, nil
 }
+
+func receiveBlockByHeight(streamer coregrpc.BlockAPI_BlockByHeightClient) (
+	*types.Block,
+	*types.BlockMeta,
+	*types.Commit,
+	*types.ValidatorSet,
+	error,
+) {
```

Comment:
Return of 5 values caught my eye. I can't find if BlockMeta returned from this method is used anywhere. If not, perhaps we can return just SignedBlock struct, which contains everything except Blockmeta?

Reply:
I am returning everything we receive in this. If you don't want block meta to be returned, I guess we could even stop sending it from core. What do you think?

Reply:
removing block meta and using signed block as a return value: 45cb70f730f7ac34a35330cb540b8d6b5ae58aac